### PR TITLE
Support Multimodal Input for Agents

### PR DESCRIPTION
**Pull request by mczhuge**
Hi, @neubig @xingyaoww @li-boxuan or any brother who are interesting,

I am currently preparing for [MLAgentBench](https://arxiv.org/abs/2310.03302) evaluation and supporting the [GAIA](https://huggingface.co/datasets/gaia-benchmark/GAIA) evaluation.

@Jiayi-Pan, please see [GPTSwarm](https://github.com/metauto-ai/gptswarm), where these multimodal readers are crucial for enhancing GAIA's performance.

**Need help with**

To maintain consistency, could someone familiar with the project use **litellm** to replace the current OpenAI API call in this file?

**Next steps**

@Jiayi-Pan, we can benchmark GAIA together. I have the datasets and a primary agent for GAIA available [here](https://github.com/metauto-ai/GPTSwarm/blob/main/swarm/environment/agents/gaia/tool_tot.py).

### CODE REVIEW

File: opendevin/core/utils/readers.py [hunk 1]
```diff
@@ -0,0 +1,538 @@
...
+        api_key = OPENAI_API_KEY
+        logger.info(f'Transcribing audio file from {file_path}.')
+        client = OpenAI(api_key=api_key)
+        try:
+            client = OpenAI()
+            with open(file_path, 'rb') as audio_file:
+                transcript = client.audio.translations.create(
+                    model='whisper-1', file=audio_file
+                )
+            return transcript.text
```
Comment (xingyaoww) [inline opendevin/core/utils/readers.py:172]: We probably need a good way to track the cost for this too...

**Review by xingyaoww — COMMENTED** [review]
I think we can actually keep a lot of these good tools (e.g., ImageReader, VideoReader) and remove some of them (e.g., zip reader - agent can just `unzip` in a `bash` shell). This is probably a good time for us to build the "agent skill library"

**Review by rbren — CHANGES_REQUESTED** [review]
See requests on structure above

**Comment by Shimada666** [comment]
tomorrow i will trying replace openai api call with litellm

**Comment by mczhuge** [comment]
> tomorrow i will trying replace openai api call with litellm

Thanks so much!

**Comment by mczhuge** [comment]
> * dependencies should go in pypackage.yml
> * this file should probably go in `opendevin/llm/readers.py`
> * we should probably make this functionality available to the LLM class somehow

Yes @rbren I am still getting more familiar with the abstraction of the whole OpenDevin system.

File: opendevin/core/utils/readers.py [hunk 2]
```diff
@@ -0,0 +1,538 @@
...
+pip install markdown
+pip install PyPDF2
+pip install openpyxl
+pip install beautifulsoup4
+pip install pylatexenc
+pip install python-pptx
+pip install xlrd
+pip install base64
+pip install opencv-python
+"""
```
Comment (yufansong) [inline opendevin/core/utils/readers.py:18]: Maybe we should move those to our dependency file.

File: opendevin/core/utils/readers.py [hunk 3]
```diff
@@ -0,0 +1,538 @@
...
+    def parse(self, file_path: Path) -> str:
+        logger.info(f'Reading ZIP file from {file_path}.')
+        file_content = ''
+        with zipfile.ZipFile(file_path, 'r') as zip_ref:
+            extract_dir = str(file_path)[:-4] + '/'
+            zip_ref.extractall(Path(extract_dir))
+            reader = FileReader()
+            for file_name in zip_ref.namelist():
+                file_content += f'File {file_name}:\n"{reader.read_file(Path(extract_dir + file_name))}"\n'
+        return file_content
```
Comment (yufansong) [inline opendevin/core/utils/readers.py:239]: What will happend if the zip file is not text data?

File: opendevin/core/utils/readers.py [hunk 4]
```diff
@@ -0,0 +1,538 @@
...
+        idx = 0
+        for base64_frame in selected_frames:
+            idx += 1
+            logger.info(
+                f'Process the {file_path}, current No. {idx * frame_interval} frame...'
+            )
+
+            api_call = self.prepare_api_call(task, base64_frame)
+            try:
+                openai_proxy: str = 'https://api.openai.com/v1/chat/completions'
```
Comment (yufansong) [inline opendevin/core/utils/readers.py:400]: Maybe also need record cost here.

File: opendevin/core/utils/readers.py [hunk 5]
```diff
@@ -0,0 +1,538 @@
...
+            'Authorization': f'Bearer {OPENAI_API_KEY}',
+        }
+
+    def parse(
+        self, file_path: Path, task: str = 'Describe this image as detail as possible.'
+    ) -> str:
+        logger.info(f'Reading image file from {file_path}.')
+
+        try:
+            openai_proxy: str = 'https://api.openai.com/v1/chat/completions'
```
Comment (yufansong) [inline opendevin/core/utils/readers.py:309]: And the cost here.

File: opendevin/core/utils/readers.py [hunk 6]
```diff
@@ -0,0 +1,538 @@
...
+# from opendevin.llm.llm import LLM
+from openai import OpenAI
+from pptx import Presentation
+from pylatexenc.latex2text import LatexNodes2Text
+
+from opendevin.core.logger import opendevin_logger as logger
+
+# TODO: Double-check where the interface for the stored API key is in OpenDevin.
+# TODO: Or ask someone familiar to change it with litellm.
+OPENAI_API_KEY = 'PUT_YOUR_OPEN_AI_API'
```
Comment (yufansong) [inline opendevin/core/utils/readers.py:50]: You can take [here](https://github.com/OpenDevin/OpenDevin/blob/8750d3e68bb24fbcc954840a3f7ff2b101981ea1/opendevin/server/agent/agent.py#L108) as a reference.

File: opendevin/core/utils/readers.py (hunk 2 above)
Comment (li-boxuan) [reply opendevin/core/utils/readers.py:18]: `pyproject.toml`, specifically

File: opendevin/core/utils/readers.py [hunk 7]
```diff
@@ -0,0 +1,538 @@
...
+    @Desc: Implementation to support reading 41 multimodal files efficiently.
+    @Ref: https://github.com/mczhuge/GPTSwarm/blob/main/swarm/environment/tools/reader/readers.py
+    """
+
+    @abstractmethod
+    def parse(self, file_path: Path) -> str:
+        """To be overriden by the descendant class"""
+
+
+class TXTReader(Reader):
```
[the remark this answers is not in the document]
Comment (li-boxuan) [reply opendevin/core/utils/readers.py:64]: I feel like (eventually) they should be micro-agents. Prompting the model about the usage of so many (and will still grow) tools would be challenging.

File: opendevin/core/utils/readers.py (hunk 4 above)
Comment (yufansong) [reply opendevin/core/utils/readers.py]: Add todo and will solve in future PR.

File: opendevin/core/utils/readers.py (hunk 5 above)
Comment (yufansong) [reply opendevin/core/utils/readers.py]: Add todo.

**Review by yufansong — APPROVED** [review]
I do some help to make sure this PR will not block other PRs progress. 
1. Move the file into `opendevin/llm/reader.py`
2. Move the installation into poetry pyproject.toml and solve the conflict with main branch.
3. For cost, we also need add cost in other parts. So prefer to solve them together in other PR.
4. For OPENAPI_KEY, I prefer to left it in the future PR. Because not sure we will shift to litellm. Don't want to write the code which may delete in the short future.

@xingyaoww @li-boxuan @rbren   WDYT

File: opendevin/core/utils/readers.py (hunk 6 above)
Comment (mczhuge) [reply opendevin/core/utils/readers.py]: I agree. I simple use the config to get the API now.

**Comment by Shimada666** [comment]
@mczhuge @yufansong 

> I do some help to make sure this PR will not block other PRs progress.我做了一些帮助，以确保这个 PR 不会阻碍其他 PR 的进展。
> 
> 1. Move the file into `opendevin/llm/reader.py`将文件移至 `opendevin/llm/reader.py`
> 2. Move the installation into poetry `pyproject.toml` and solve the conflict with main branch.将装置移入诗歌 `pyproject.toml` ，解决与主分支的冲突。
> 3. For cost, we also need add cost in other parts. So prefer to solve them together in other PR.对于成本，我们还需要在其他部分增加成本。所以更愿意在其他 PR 中一起解决它们。
> 4. May discuss whether we will shift to litellm可能会讨论我们是否会转向 litellm
> 
> @xingyaoww @li-boxuan @rbren WDYT WDYT公司

Oh, I've done some work to replace the OpenAI API call with litellm, and I've also done some code optimizations. Please review them, and if litellm is confirmed to be needed, I will submit my code after this PR is merged.

**Comment by mczhuge** [comment]
@rbren Seems this PR is ready, and @Shimada666 will further improve after merging.

File: opendevin/llm/readers.py [hunk 8]
```diff
@@ -0,0 +1,523 @@
...
+import zipfile
+from abc import ABC, abstractmethod
+from pathlib import Path
+from typing import Any, Union
+
+import charset_normalizer
+import cv2
+import docx
+
+# import markdown
```
Comment (li-boxuan) [inline opendevin/llm/readers.py:16]: Dead code

```suggestion
```

File: opendevin/llm/readers.py [hunk 9]
```diff
@@ -0,0 +1,523 @@
...
+import requests
+import yaml
+from bs4 import BeautifulSoup
+
+# from litellm import completion as litellm_completion
+from openai import OpenAI
+from pptx import Presentation
+from pylatexenc.latex2text import LatexNodes2Text
+
+# from opendevin.llm.llm import LLM
```
Comment (li-boxuan) [inline opendevin/llm/readers.py:29]:
```suggestion
```

File: opendevin/llm/readers.py [hunk 10]
```diff
@@ -0,0 +1,523 @@
...
+from opendevin.core.logger import opendevin_logger as logger
+
+# TODO: Find way to directly get the API key from ConfigType.LLM_API_KEY or change it with litellm.
+OPENAI_API_KEY = config.llm.api_key
+
+
+class Reader(ABC):
+    """
+    @Desc: Implementation to support reading 41 multimodal files efficiently.
+    @Ref: https://github.com/mczhuge/GPTSwarm/blob/main/swarm/environment/tools/reader/readers.py
```
Comment (li-boxuan) [inline opendevin/llm/readers.py:40]: deadlink

```suggestion
    @Ref: https://github.com/metauto-ai/GPTSwarm/blob/main/swarm/environment/tools/reader/readers.py
```

**Review by li-boxuan — COMMENTED** [review]
This code will reside in our core package, we really need some unit tests. Otherwise our test coverage for the entire project will drop by 18%... It shouldn't be hard to do so for many readers here. I am okay with doing it later as long as we do it eventually...

I added an unit test as an example.

File: opendevin/llm/readers.py (hunk 8 above)
Comment (li-boxuan) [reply opendevin/llm/readers.py:16]: nvm I see markdown might still be useful. It looks like it's just temporarily commented out.

File: poetry.lock [hunk 11]
```diff
@@ -5272,13 +5394,13 @@ files = [
 
 [[package]]
 name = "requests"
-version = "2.31.0"
+version = "2.32.1"
```
Comment (li-boxuan) [inline poetry.lock:5397]: This has caused integration tests to fail... 
see https://github.com/docker/docker-py/issues/3256

I'll revert this change

Comment (li-boxuan) [reply poetry.lock]: Rule of thumb:

Do

```bash
poetry lock --no-update
```

Don't do

```bash
poetry lock
```

Comment (yufansong) [reply poetry.lock]: My bad, I should try `poetry lock --no-update`

**Comment by Shimada666** [comment]
It seems there are some conflicts that need to be resolved.
I hope this PR can be merged soon. I'm eager to contribute! 😉

**Comment by mczhuge** [comment]
> It seems there are some conflicts that need to be resolved.
> I hope this PR can be merged soon. I'm eager to contribute! 😉

Me too. Hope this PR could be merged soon.

**Comment by mczhuge** [comment]
> It seems there are some conflicts that need to be resolved.
> I hope this PR can be merged soon. I'm eager to contribute! 😉

Solved the conflicts.

**Review by mczhuge — COMMENTED** [review]
Looks good now.

**Review by rbren — CHANGES_REQUESTED** [review]
This is really neat functionality! But we need to think through this a bit more thoroughly.

It looks like these readers are reading from paths on the local disk, where the API server is running. They should really be reading files via [the runtime](https://github.com/OpenDevin/OpenDevin/tree/main/opendevin/runtime), which has access to the user's workspace. Otherwise we'll end up with problems when e.g. the user is using E2B or another remote runtime. Not to mentions security concerns.

For the multimodal cases (e.g. images) we'll need to figure out a better way to stick their output into the EventStream (maybe just the b64 image?), and how to let the agent pass the data to the LLM (which you're getting at with `prepare_api_call`.

I'm imagining a flow like this:
* Agent sends a `readFile(path="foo.pdf")` action
* We read the bytes of the file from the runtime
* We do any post-processing of the bytes using these readers (e.g. extracting PDF text) 
* We put the resulting FileReadObservation into the event stream

If there's a `FileReadObservation` for e.g. a PNG image, the agent would be responsible for doing the work currently in `prepare_api_call`--we should probably put helper functions in the LLM class (e.g. `get_message_for_image(b64: string)`)

**Comment by neubig** [comment]
> we'll need to figure out a better way to stick their output into the EventStream (maybe just the b64 image?)

Note that this is already what we do with browser screenshots I believe.

**Comment by Jiayi-Pan** [comment]
This is awesome! This looks like a good starting point to get multi-modal understanding / gaia agent started.

Kind of horizontal to this, I do wonder if we should also consider a general approach in the long run (see discussions #1911 between @frankxu2004 @li-boxuan and myself).

Basically, once we enable the agent to directly read browser's pixel observation space, it indirectly solves image, video, pdf understanding, since agent can just open the file in the browser and read it.

For other files like `docx`, `json`, since our agent is capable of installing pacakages and have a jupyter repl interface, it can automatically figure out how to do so.

> One interesting thing I discovered during testing a DOCX understanding question is that Open-Dev’s agent has a sufficiently broad action space allowing the agent to develop multi-modal understanding skill by itself.
The agent decided to first install the python-docx package and use it within Jupyter to assist with understanding the docx document.

**Comment by xingyaoww** [comment]
@Jiayi-Pan agreed! For the time being (i mean for Neurips submission) we can assume agent can only perceive multi-modal via tools (e.g., defined in this PR). But as a natural next step, we should allow the agent to directly perceive raw pixels from browser, which would "solves image, video, pdf understanding, since agent can just open the file in the browser and read it." as you described.

**Comment by frankxu2004** [comment]
Agreed. In the long run we can probably eventually rely on OS or browser interface to handle these observations (audio capture, anyone?), without having to hand craft too much about tools. But to make things practical, creating these programmatic tools are useful for now!

**Comment by mczhuge** [comment]
Considering the current situation I think it is a good starting point to first use current reader and link them into ``action`` and ``observation`` and then *improve these functions iteratively*. 

Considering the concept of a 'pixel observation space', I believe it can become feasible with the release of GPT-5 (or more powerful LLMs). Currently, translating any modality simply by observing can lead to uncontrollable hallucinations and loss of information.

**Comment by neubig** [comment]
Thanks for all the discussion everyone! Do we know what the next steps are? Asking just because it's kinda time sensitive.

**Comment by Shimada666** [comment]
@neubig 
As @xingyaoww mentioned, this is a good opportunity to build the agent skill library, so I have integrated the capabilities of some readers into agent skill as part of #2016 . However, the decision on which readers to retain still requires input from everyone. I hope to get everyone's suggestion!

**Comment by xingyaoww** [comment]
Thanks @mczhuge for contributing these multi-modal readers! https://github.com/OpenDevin/OpenDevin/pull/2016 refactored those and included them into `agentskills` -- let's build on top of these tools for a strong GAIA agent :)

**Pull request closed without merging**